design/ceph-nfs-ganesha.md

```diff
@@ -0,0 +1,190 @@
+storage backend for rook (the general NFS solution) in the following ways:
+
+* It will use the rook's Ceph operator and not a separate NFS operator to
+  deploy the NFS server pods.
+
+* The NFS server pods will be directly configured with CephFS or RGW
+  backend setup by rook, and will not require CephFS or RGW to be mounted
+  in the NFS server pod with a PVC.
+
+## Design of Ceph NFS-Ganesha CRD
+
+The NFS-Ganesha server and its exports' settings will be exposed to rook as a
+Custom Resource Definition (CRD). Creating the nfs-ganesha CRD will launch
+NFS Ganesha server pods that will be configured with a rook CephFS filesystem
+or a RGW object store backend. The servers' exports will be stored in
+[k8s ConfigMaps]; the NFS client recovery data will be stored in a Ceph RADOS
+pool; and the servers will have stable IP addresses by using [k8s Service].
+This allows the servers pods to be scalable and highly available.
+
+### Prerequisities
+
+- A running rook Ceph filesystem or object store, whose namespaces will be
+  exported by the Ganesha servers.
+  e.g.,
+  ```
+  kubectl create -f filesystem.yaml
+  ```
+
+- An existing RADOS pool (e.g., CephFS's data pool) or a pool created with a
+  [Ceph Pool CRD] to store NFS client recovery data.
+
+
+### Ceph NFS-Ganesha CRD
+
+The NFS-Ganesha CRD will specify the following:
+
+- Number of active Ganesha servers
+
+- Placement of the Ganesha servers
+
+- Resource limits (memory, CPU) of the Ganesha pods
+
+- Export settings
+
+- RADOS pool and namespace where the NFS client recovery data will be stored
+
+
+Below is an example NFS-Ganesha CRD, `nfs-ganesha.yaml`
+
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: NFSGanesha
+metadata:
+  # The name of Ganesha server cluster to create, which will be reflected in
+  # the name(s) of the ganesha server pod(s)
+  name: mynfs
+  # The namespace of the Rook cluster where the Ganesha server cluster is
+  # created.
+  namespace: rook-ceph
+spec:
+  # The Ceph filesystem or RGW that will be configured as the Ganesha server's
+  # backend. It will be the identified by the name of the filesytem or object
+  # store CRD.
+  backendName: myfs
+  ganeshaServer:
+    # the number of active ganesha servers
+    active: 3
+    # where to run the nfs ganesha server
+    placement:
+    #  nodeAffinity:
+    #    requiredDuringSchedulingIgnoredDuringExecution:
+    #      nodeSelectorTerms:
+    #      - matchExpressions:
+    #        - key: role
+    #          operator: In
+    #          values:
+    #          - ganesha-node
+    #  tolerations:
+    #  - key: ganesha-node
+    #    operator: Exists
+    #  podAffinity:
+    #  podAntiAffinity:
+    # The requests and limits set here allow the ganesha pod(s) to use half
+    # of one CPU core and 1 gigabyte of memory
+    resources:
+    #  limits:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+    #  requests:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+
+  # Export settings
+  exports:
```

Comment:
Questions about the exports:
- This is a list of exports and at least one is required, correct? 
- `allowedClients` is a list? Do we need the `clients` under this? 
- The allowed clients can be a mix of ip ranges and host names?
- Are `accessMode` and `squash` settings for the export, or just for the `allowedClients`?

Reply:
> This is a list of exports and at least one is required, correct?
If no exports are defined in ganesha.conf, the ganesha server will still start but won't export anything. 

> allowedClients is a list? Do we need the clients under this?
Yes it's a list. There can be clients lists (client blocks in nfs-ganesha project parlance ) with different access permissions and root squash permissions. See this,
https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/config_samples/export.txt#L175

> The allowed clients can be a mix of ip ranges and host names?
Yes.


>  Are accessMode and squash settings for the export, or just for the allowedClients?
They can be for the entire export, and can be per client block/group as mentioned earlier.
See,
https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/config_samples/export.txt#L153
and
https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/config_samples/export.txt#L175